src/vos/vos_space.c

```diff
@@ -0,0 +1,344 @@
+	/* Assume one more key tree node created */
+	size += umem_slab_usize(umm, VOS_SLAB_KEY_NODE);
+
+	return size;
+}
+
+/*
+ * Estimate how much space will be consumed by an update request. This
+ * conservative esimation always assumes new object, dkey, akey will be
+ * created for the update.
+ */
+static void
+estimate_space(struct vos_pool *pool, daos_key_t *dkey, unsigned int iod_nr,
+	       daos_iod_t *iods, struct dcs_iod_csums *iods_csums,
+	       daos_size_t *space_est)
+{
+	struct umem_instance	*umm = vos_pool2umm(pool);
+	struct dcs_csum_info	*csums, *recx_csum;
+	daos_iod_t		*iod;
+	daos_recx_t		*recx;
+	uint16_t		 media;
+	daos_size_t		 size, scm, nvme = 0 /* in blk */;
+	int			 i, j;
+
+	/* Object record */
+	scm = umem_slab_usize(umm, VOS_SLAB_OBJ_DF);
+	/* Assume one more object tree node created */
+	scm += umem_slab_usize(umm, VOS_SLAB_OBJ_NODE);
+
+	/* Dkey */
+	scm += estimate_space_key(umm, dkey);
+
+	for (i = 0; i < iod_nr; i++) {
+		iod = &iods[i];
+
+		/* Akey */
+		scm += estimate_space_key(umm, &iod->iod_name);
+
+		csums = iods_csums ? iods_csums[i].ic_data : NULL;
+		/* Single value */
+		if (iod->iod_type == DAOS_IOD_SINGLE) {
+			size = iod->iod_size;
+			media = vos_media_select(pool, iod->iod_type, size);
+
+			/* Single value record */
+			if (media == DAOS_MEDIA_SCM) {
+				scm += vos_recx2irec_size(size, csums);
+			} else {
+				scm += vos_recx2irec_size(0, csums);
+				if (iod->iod_size != 0)
+					nvme += vos_byte2blkcnt(iod->iod_size);
+			}
+			/* Assume one more SV tree node created */
+			scm += umem_slab_usize(umm, VOS_SLAB_SV_NODE);
+			continue;
+		}
+
+		/* Array value */
+		for (j = 0; j < iod->iod_nr; j++) {
+			recx = &iod->iod_recxs[j];
+			recx_csum = csums ? &csums[j] : NULL;
+
+			size = recx->rx_nr * iod->iod_size;
+			media = vos_media_select(pool, iod->iod_type, size);
+
+			/* Extent */
+			if (media == DAOS_MEDIA_SCM)
+				scm += size;
+			else if (size != 0)
+				nvme += vos_byte2blkcnt(size);
+			/* EVT desc */
+			scm += umem_slab_usize(umm, VOS_SLAB_EVT_DESC);
+			/* Checksum */
+			scm += recx_csum_len(recx, recx_csum, iod->iod_size);
+			/* Assume one more evtree node created */
+			scm += umem_slab_usize(umm, VOS_SLAB_EVT_NODE);
+		}
+	}
+
+	space_est[DAOS_MEDIA_SCM] = scm;
+	space_est[DAOS_MEDIA_NVME] = nvme * VOS_BLK_SZ;
+}
+
+int
+vos_space_hold(struct vos_pool *pool, daos_key_t *dkey, unsigned int iod_nr,
+	       daos_iod_t *iods, struct dcs_iod_csums *iods_csums,
+	       daos_size_t *space_hld)
+{
+	struct vos_pool_space	vps = { 0 };
+	daos_size_t		space_est[DAOS_MEDIA_MAX] = { 0, 0 };
+	daos_size_t		scm_left, nvme_left;
+	int			rc;
+
+	rc = vos_space_query(pool, &vps, false);
+	if (rc) {
+		D_ERROR("Query pool:"DF_UUID" space failed. "DF_RC"\n",
+			DP_UUID(pool->vp_id), DP_RC(rc));
+		return rc;
+	}
+
+	scm_left = SCM_FREE(&vps);
+	if (scm_left < SCM_SYS(&vps))
+		goto error;
+
+	scm_left -= SCM_SYS(&vps);
+	if (scm_left < POOL_SCM_HELD(pool))
```

Comment:
(just a question) similarly, is it ever expected that SCM_FREE(&vps) - SCM_SYS(&vps) could be less than POOL_SCM_HELD(pool)?

Reply:
Yes, as I commented above, when GC or aggregation temporarily consumed some space, the SCM_FREE() could be less than 'total - used - reserve_sys - reserve_held'.